10-JavaScript/arrow-functions.md

```diff
@@ -0,0 +1,261 @@
+
+Here is an example for defining a function with no parameters which returns a string "hello world".  Convert this into an arrow function and minimize the code typed.
+
+```javascript
+const hello = function() {
+  return "hello world"
+};
+```
+
+### Arrow Functions with parameters
+
+```javascript
+const squarer = (number) => {
+  return number * number
+}
+```
+This arrow function takes a single number and returns that number squared.  Like normal functions arrow functions have their parameters listed in the parentheses.  
+
+So this is a pretty compact syntax, but... we're not done.  Because this function takes a **single** argument, we can dispense with the parentheses around `number`.
+
+```javascript
+const squarer = number => {return number * number;};
+squared([3, 4, 5]);
+> [ 9, 16, 25 ]
+```
+
+**Note**: This works **only** for functions with a single argument.  Zero parameters or more than one will require you to include the parentheses.  
+
+For example with two parameters
+
+```javascript
+const add = (x, y) => x + y;
+```
+
+Just like earlier since `squarer` has a 1-line function body we can remove the curly braces.
+
+```javascript
+const squarer = number => return number * number; ;
+squared([3, 4, 5]);
+> [ 9, 16, 25 ]
+```
+
+## Inline Arrow Functions Returning Objects
+
+If your arrow function returns a JavaScript object like this normal function.
+
+```javascript
+const makePerson = function(id, name) {
+  return {id: id,
+    name: name
+  };
+}
+```
+
+You would need to surround the returned object with parentheses like this:
+
+```javascript
+const makePerson = (id, name) => ({ id: id, name: name });
+```
+
+## As An Anonymous Function
+
+Arrow functions work well with collection functions like `map`, `select`, and `sort`.
+
+```javascript
+let numbers = [3, 4, 5, 6];
+
+let doubled = numbers.map(function(num) {
+  return num + num;  
+});
+
+// as an arrow function
+
+doubled = numbers.map( num => num + num);
+```
+
+In the above example we map an array to one where each element is doubled.  Looking at the two examples the arrow function, once you adjust to the new syntax, is more compact and visually conveys the concept that we're mapping each number (`num`) to twice it's value.
+
+### More Reason to use Arrow Functions
+
+There are additional benefits to using arrow functions beyond minimized syntax.  
+
+Lets look at a more complex example:
+
+```javascript
+class Dog {
+  constructor(name, age) {
+    this.name = name;
+    this.age = age;
+    this.interval = setInterval(function() {
+      console.log(this.age);
+      this.age++;
+      if (this.age > 10) clearInterval(this);
+
+    }, 1000);
+  }
+}
+
+let fido = new Dog("Fido", 3);
+```
+
+The code above is **supposed** to make the dog 1 year older each second.  Instead it prints `NaN` over and over again.  The reason is that every traditional function gets it's own instance of `this`.  So the anonymous function does not know about Dog's instance variables.
```

Comment:
I like the idea of this example, but with the asynchronous-ness I worry that it'll be a little over their heads at this point. Here is a simpler version:

```js
obj = {
  thing: 1,
  increment: function() {
    [1,2,3].forEach(function(num) {
      this.thing += num;
    });
  }
}

console.log(obj.thing);
obj.increment();
console.log(obj.thing); // Expected 7, but prints 1
```